src/internal/connector/exchange/service_iterators.go

```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
 	oldDeltas map[string]string,
 ) error {
 	var (
-		errs           error
-		collectionType = CategoryToOptionIdentifier(qp.Category)
+		errs error
+		oi   = CategoryToOptionIdentifier(qp.Category)
 		// folder ID -> delta url for folder.
 		deltaURLs = map[string]string{}
+		prevPaths = map[string]string{}
 	)
 
 	for _, c := range resolver.Items() {
+		if qp.FailFast && errs != nil {
+			return errs
+		}
+
 		dirPath, ok := pathAndMatch(qp, c, scope)
 		if !ok {
 			continue
 		}
 
+		cID := *c.GetId()
+
 		// Create only those that match
 		service, err := createService(qp.Credentials, qp.FailFast)
 		if err != nil {
-			errs = support.WrapAndAppend(
-				qp.ResourceOwner+" FilterContainerAndFillCollection",
-				err,
-				errs)
-
-			if qp.FailFast {
-				return errs
-			}
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+			continue
 		}
 
 		edc := NewCollection(
 			qp.ResourceOwner,
 			dirPath,
-			collectionType,
+			oi,
 			service,
 			statusUpdater,
 		)
-		collections[*c.GetId()] = &edc
+		collections[cID] = &edc
 
 		fetchFunc, err := getFetchIDFunc(qp.Category)
 		if err != nil {
-			errs = support.WrapAndAppend(
-				qp.ResourceOwner,
-				err,
-				errs)
-
-			if qp.FailFast {
-				return errs
-			}
-
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
 			continue
 		}
 
-		dirID := *c.GetId()
-		oldDelta := oldDeltas[dirID]
-
-		jobs, delta, err := fetchFunc(ctx, edc.service, qp.ResourceOwner, dirID, oldDelta)
+		jobs, delta, err := fetchFunc(ctx, edc.service, qp.ResourceOwner, cID, oldDeltas[cID])
 		if err != nil {
-			errs = support.WrapAndAppend(
-				qp.ResourceOwner,
-				err,
-				errs,
-			)
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
 		}
 
 		edc.jobs = append(edc.jobs, jobs...)
 
 		if len(delta) > 0 {
-			deltaURLs[dirID] = delta
+			deltaURLs[cID] = delta
+		}
+
+		// add the current path for the container ID to be used in the next backup
+		// as the "previous path", for reference in case of a rename or relocation.
+		if itp, err := resolver.IDToPath(ctx, cID); err != nil {
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+		} else {
+			// Some root folders have no display name. The exchange path will
+			// fail without that, so we inject a fallback here just for
+			// the previousPaths handling.
+			els := itp.Elements()
+			if len(els) == 0 || len(els[len(els)-1]) == 0 {
+				itp = itp.Append(rootFolderAlias)
+			}
+
+			p, err := itp.ToDataLayerExchangePathForCategory(
+				qp.Credentials.AzureTenantID,
+				qp.ResourceOwner,
+				qp.Category,
+				false)
+			if err != nil {
+				errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+			} else {
+				prevPaths[cID] = p.Folder()
```

Comment:
I was thinking about this, and that we'd probably be safer overall storing slice over the folder as a string.  Saves us from parsing collisions like this.  Something I'll change in a follow-up, to not overwhelm this PR.

Reply:
>safer overall storing slice over the folder as a string.

I'm not sure I understand this. Can you explain more or point me to the code when you have it if it's going to be in this PR

Reply:
Ah, sorry, late night brain was getting to me.  I did half a read and then misinterpreted your comment as worrying about string->path parsing.  The test is a good idea, and I'll add it in, but personally I have very few worries about json decoding/encoding. Or, at least, the golang package for it.  If graph has its own bugs there I'm not aware of them.

Reply:
mmm I don't think it would be a bug in the golang JSON parser. I'm more worried about us not properly escaping things we need to when handing data to it and that causing problems down the line. It's more of a misunderstanding of the API and assumptions than a bug

Reply:
Topic resolved in sync discussion; appropriate tests added.